File: Search-Spaces/search_spaces.js
// Lookup Spaces by ID
// https://developer.twitter.com/en/docs/twitter-api/spaces/lookup

const needle = require('needle');

// The code below sets the bearer token from your environment variables
// To set environment variables on macOS or Linux, run the export command below from the terminal:
// export BEARER_TOKEN='YOUR-TOKEN'
const token = process.env.BEARER_TOKEN;

const endpointUrl = `https://api.twitter.com/2/spaces/search`;

async function getRequest() {

    // Edit query parameters below and specify a search query
    // optional params: host_ids,conversation_controls,created_at,creator_id,id,invited_user_ids,is_ticketed,lang,media_key,participants,scheduled_start,speaker_ids,started_at,state,title,updated_at
    const params = {
        'query': 'NBA', // Replace the value with your search term
        'space.fields': 'title,created_at',
        'expansions': 'creator_id'
    }

    const res = await needle('get', endpointUrl, params, {
        headers: {
            "User-Agent": "v2SpacesSearchJS",
            "authorization": `Bearer ${token}`
        }
    })

    if (res.body) {
        return res.body;
    } else {
        throw new Error('Unsuccessful request');
    }
}

(async () => {

    try {
        // Make request
        const response = await getRequest();
        console.dir(response, {
            depth: null
        });

    } catch (e) {
        console.log(e);
        process.exit(-1);
    }
    process.exit();
})();

File: Search-Spaces/search_spaces.rb
# This script uses your bearer token to authenticate and make a Spaces lookup request

require 'json'
require 'typhoeus'

# The code below sets the bearer token from your environment variables
# To set environment variables on macOS, run the export command below from the terminal:
# export BEARER_TOKEN='YOUR-TOKEN'
bearer_token = ENV["BEARER_TOKEN"]

# Endpoint URL
spaces_search_url = "https://api.twitter.com/2/spaces/search"

# Replace this value with your search term
query = "NBA"

# Add or remove parameters below to adjust the query and response fields within the payload
# See docs for list of param options: https://developer.twitter.com/en/docs/twitter-api/spaces/search/api-reference
query_params = {
  "query": query, # Required
  "space.fields": 'title,created_at', 
  'expansions': 'creator_id'
}

def get_space(url, bearer_token, query_params)
  options = {
    method: 'get',
    headers: {
      "User-Agent": "v2SpacesSearchRuby",
      "Authorization": "Bearer #{bearer_token}"
    },
    params: query_params
  }

  request = Typhoeus::Request.new(url, options)
  response = request.run

  return response
end

response = get_space(spaces_search_url, bearer_token, query_params)
puts response.code, JSON.pretty_generate(JSON.parse(response.body))
